File: android/src/main/java/com/dgismaps/DGisSdkHolder.kt
package com.dgismaps

import android.content.Context
import ru.dgis.sdk.DGis
import ru.dgis.sdk.PersonalDataCollectionConsent
import ru.dgis.sdk.platform.HttpOptions
import ru.dgis.sdk.platform.KeyFromAsset
import ru.dgis.sdk.platform.KeyFromString
import ru.dgis.sdk.platform.KeySource
import ru.dgis.sdk.platform.LogLevel
import ru.dgis.sdk.platform.LogOptions
import ru.dgis.sdk.Context as DGisContext

object DGisSdkHolder {
  private val lock = Any()
  private var sdkContext: DGisContext? = null
  private var initializedKey: String? = null

  val isInitialized: Boolean
    get() = synchronized(lock) { sdkContext != null }

  fun initialize(context: Context, apiKey: String, logLevel: String?): Boolean = synchronized(lock) {
    val existingKey = initializedKey
    if (existingKey != null) {
      if (existingKey == apiKey) {
        return@synchronized true
      }
      throw IllegalStateException("2GIS SDK is already initialized with another key")
    }

    val level = when (logLevel) {
      "verbose" -> LogLevel.VERBOSE
      "warning" -> LogLevel.WARNING
      "error" -> LogLevel.ERROR
      "off" -> LogLevel.OFF
      else -> LogLevel.INFO
    }

    val keySource = resolveKeySource(context, apiKey)

    sdkContext = DGis.initialize(
      appContext = context.applicationContext,
      httpOptions = HttpOptions(useCache = true),
      logOptions = LogOptions(systemLevel = level),
      keySource = keySource,
      dataCollectConsent = PersonalDataCollectionConsent.GRANTED
    )
    initializedKey = apiKey
    true
  }

  private fun resolveKeySource(context: Context, apiKey: String): KeySource {
    // Prefer the signed binary key file shipped as an asset.
    val assets = runCatching { context.applicationContext.assets.list("") ?: emptyArray<String>() }
      .getOrDefault(emptyArray())
    if (assets.contains("dgissdk.key")) {
      return KeySource(KeyFromAsset("dgissdk.key"))
    }
    return KeySource(KeyFromString(apiKey))
  }

  fun requireContext(): DGisContext = synchronized(lock) {
    sdkContext ?: throw IllegalStateException(
      "2GIS SDK is not initialized. Call DgisMapsModule.initialize(...) before mounting DgisMapsView."
    )
  }

  fun contextOrNull(): DGisContext? = synchronized(lock) {
    sdkContext
  }
}

File: ios/DGisSdkHolder.swift
import DGis
import Foundation

@objc(DGisSdkHolder)
public final class DGisSdkHolder: NSObject {
  @objc public static let shared = DGisSdkHolder()

  private let lock = NSLock()
  private var container: DGis.Container?
  private var initializedApiKey: String?

  private override init() {
    super.init()
  }

  @objc public var isInitialized: Bool {
    lock.lock()
    defer { lock.unlock() }
    return container != nil
  }

  @objc public func initialize(_ apiKey: String, logLevel: String?) throws {
    lock.lock()
    defer { lock.unlock() }

    if let initializedApiKey {
      if initializedApiKey == apiKey {
        return
      }

      throw NSError(
        domain: "DgisMaps",
        code: 1001,
        userInfo: [NSLocalizedDescriptionKey: "2GIS SDK is already initialized with another key"]
      )
    }

    let level: DGis.LogLevel
    switch logLevel {
    case "verbose":
      level = .verbose
    case "warning":
      level = .warning
    case "error":
      level = .error
    case "off":
      level = .off
    default:
      level = .info
    }

    container = DGis.Container(
      keySource: Self.resolveKeySource(apiKey: apiKey),
      logOptions: LogOptions(systemLevel: level),
      httpOptions: HttpOptions(),
      personalDataCollectionOptions: PersonalDataCollectionOptions(
        personalDataCollectionConsent: .granted
      )
    )
    initializedApiKey = apiKey
  }

  private static func resolveKeySource(apiKey: String) -> KeySource {
    // Mobile SDK validates the signed binary dgissdk.key, not the API string —
    // KeyFromString is left only as a dev-time fallback so init() doesn't throw
    // synchronously, but the vendor will reject it with "File with key info is
    // invalid". A consumer who hits this warning needs to add dgissdk.key to
    // the app target's Copy Bundle Resources phase.
    if let path = Bundle.main.path(forResource: "dgissdk", ofType: "key") {
      return .fromFile(KeyFromFile(path: path))
    }
    NSLog("[DGisSdkHolder] WARNING: dgissdk.key missing from app bundle — falling back to KeyFromString which the Mobile SDK will reject.")
    return .fromString(KeyFromString(contents: apiKey))
  }

  public func sdkOrNil() -> DGis.Container? {
    lock.lock()
    defer { lock.unlock() }
    return container
  }

  public func requireSdk() throws -> DGis.Container {
    lock.lock()
    defer { lock.unlock() }

    guard let container else {
      throw NSError(
        domain: "DgisMaps",
        code: 1000,
        userInfo: [NSLocalizedDescriptionKey: "2GIS SDK is not initialized. Call DgisMapsModule.initialize(...) before mounting DgisMapsView."]
      )
    }

    return container
  }
}
